src/test/regress/expected/multi_deparse_function.out

```diff
@@ -0,0 +1,311 @@
+--
+-- Regression tests for deparsing ALTER/DROP FUNCTION Queries
+--
+-- This test implements all the possible queries as of Postgres 11:
+-- 
+-- ALTER FUNCTION name [ ( [ [ argmode ] [ argname ] argtype [, ...] ] ) ]
+--     action [ ... ] [ RESTRICT ]
+-- ALTER FUNCTION name [ ( [ [ argmode ] [ argname ] argtype [, ...] ] ) ]
+--     RENAME TO new_name
+-- ALTER FUNCTION name [ ( [ [ argmode ] [ argname ] argtype [, ...] ] ) ]
+--     OWNER TO { new_owner | CURRENT_USER | SESSION_USER }
+-- ALTER FUNCTION name [ ( [ [ argmode ] [ argname ] argtype [, ...] ] ) ]
+--     SET SCHEMA new_schema
+-- ALTER FUNCTION name [ ( [ [ argmode ] [ argname ] argtype [, ...] ] ) ]
+--     DEPENDS ON EXTENSION extension_name
+-- 
+-- where action is one of:
+-- 
+--     CALLED ON NULL INPUT | RETURNS NULL ON NULL INPUT | STRICT
+--     IMMUTABLE | STABLE | VOLATILE | [ NOT ] LEAKPROOF
+--     [ EXTERNAL ] SECURITY INVOKER | [ EXTERNAL ] SECURITY DEFINER
+--     PARALLEL { UNSAFE | RESTRICTED | SAFE }
+--     COST execution_cost
+--     ROWS result_rows
+--     SET configuration_parameter { TO | = } { value | DEFAULT }
+--     SET configuration_parameter FROM CURRENT
+--     RESET configuration_parameter
+--     RESET ALL
+-- 
+-- DROP FUNCTION [ IF EXISTS ] name [ ( [ [ argmode ] [ argname ] argtype [, ...] ] ) ] [, ...]
+--     [ CASCADE | RESTRICT ]
+SET citus.next_shard_id TO 20020000;
+CREATE SCHEMA function_tests;
+SET search_path TO function_tests;
+SET citus.shard_count TO 4;
+SET client_min_messages TO INFO;
+CREATE FUNCTION deparse_test(text)
+	RETURNS text
+	AS 'citus'
+ 	LANGUAGE C STRICT;
+-- Create a simple function
+CREATE FUNCTION add(integer, integer) RETURNS integer
+    AS 'select $1 + $2;'
+    LANGUAGE SQL
+    IMMUTABLE
+    RETURNS NULL ON NULL INPUT;
+SELECT deparse_test($cmd$
+ALTER FUNCTION  add CALLED ON NULL INPUT
+$cmd$);
+                       deparse_test                        
+-----------------------------------------------------------
+ ALTER FUNCTION function_tests.add() CALLED ON NULL INPUT;
```

Comment:
one problem across all the tests is that the deparsed query cannot be executed:
```SQL
ALTER FUNCTION function_tests.add() CALLED ON NULL INPUT;
ERROR:  function function_tests.add() does not exist
Time: 0.612 ms
```
Which should not happen because we're planning to execute this queries on the workers.

So, we should add the parameter types to the output. 

And, more importantly, maybe let `deparse_test ` to connect to the node, and execute the command so that we'd not have any such cases. Otherwise, you'll need to manually ensure that the commands work fine

Reply:
Connecting to the workers and executing the commands is best done in an end to end fashion by hooking the deparser to the utility hook as it done in (#2995, #2994, #2993, #2992, #2991). Doing so in this test seems redundant and a lot of work.

For now I think it is our job as author and reviewers to critically inspect the output of the tests.

Reply:
> For now I think it is our job as author and reviewers to critically inspect the output of the tests.

Why not let Postgres do the testing for us :)  I don't think that it is a lot of work, simply call into StandardProcessUtility.  Besides, testing the deparsing logic here would help us to keep scope the smaller on the mentioned PRs. I'm not saying that we should not test end-to-end in those PRs, but I cannot understand the pushback that'd prevent us to be confident with deparsing in this PR.

Anyway, I'll leave it to you guys.

Reply:
First of all, I now see the initial problem you have pointed out here, it got overshadowed by the suggestion to 'connect' to a node to execute it. @hanefi As onder pointed out the deparsing logic has a flaw which causes it to fail on this statement.
`add` and `add()` have different meanings in postgres. The first would resolve to a function as long as there is only 1 named function `add`. It ignores the input arguments as it can unambiguously resolve `add` to `add(int,int)`. However the deparsed function name has the parenthesis so postgres assumes you would be looking for a function `add` with no arguments which is not what we want. Can you make sure the deparser doesn't output parenthesis for these kinds of deparses? Make sure to also add a test that would target a function without arguments as the parenthesis would need to be added if it has been overloaded. eg.
```sql
magic_function()
magic_function(int,int)

ALTER FUNCTION magic_function ... -- should deparse without parenthesis, even though it will fail during execution
ALTER FUNCTION magic_function() ... -- sholuld deparse with an empty argument list
ALTER FUNCTION magic_function(int, int) ... -- should deparse with its full list of arguments
```

@onderkalaci On the point of executing the commands. Although the idea of executing the commands locally is appealing, more so than connecting to a worker as I understood from your initial message, however it is not as clear cut as it seems.

Executing the commands have side effects on the target function that we would not want to have during the deparsing tests. This could be prevented by executing the command in a transaction that rolls back after it confirmed the change applied correctly.

The definition of correctly is hard. In the bug you pointed out here it would have sufficed to assert no error has been raised by executing the command locally. Given postgres would not be able to find `add()` would cause the exception to be raised. Other test cases explicitly test non-existing schema's and non-existing function names to verify those can be deparsed as well. Executing these statements locally would always fail, which is desired behaviour.

One could argue we could create a savepoint, execute the original command locally, record its output, restore the savepoint, execute the deparsed statement verify the output is the same. Now we are getting somewhere, but still it will not catch subtile issues.

To correctly test the correct execution of the statement we now need to also inspect all the side effects; did the deparsed statement address the exact same object, did it apply the same changes, etc. To get to a hardened version of this that would work throughout the test suite is a though job which I don't think we could afford at the moment.

Reply:
In a commit I pushed yesterday (717b41f53fbd5645cdcfd6fa3f8c407525bf84c5) I resolved the erronous `add()` pieces in the tests.

I will however add some test cases with several functions with the same name but with a different list of parameters to make sure that we address the same function in the workers as well.

I have also created a utility SQL function that I can possibly use to test that the queries are able to run on the workers.

```SQL
CREATE FUNCTION deparse_and_run_on_workers(text)
    RETURNS SETOF record AS $fnc$
    	WITH deparsed_query AS ( SELECT deparse_test($1) qualified_query )
    	SELECT run_command_on_workers(qualified_query) FROM deparsed_query 
    $fnc$ LANGUAGE SQL;
```

Even though I do not check that the resulting changes in the worker are the same as that in the coordinator, I think seeing that the deparsed query can successfully run on the worker should be enough for now.

Reply:
actually they don't have to be run on the worker, you can execute them locally.
multiple renames, and changes of schema's are a bit tricky if you apply them, either locally or remote. Hence my suggestion to create a savepoint to rollback to. This will execute the query but not persist the side effects, meaning subsequent calls can all work on the same source.

Reply:
Executing them on the workers has been implemented